opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -214,8 +226,264 @@ namespace Opm {
+            //wellModel().recoverWellSolutionAndUpdateWellState(x);
+            std::cout << "Printing matrix residual in backward mode" << std::endl;
+            std::cout << "Inf norm " << ebosResid.infinity_norm() << std::endl;
+            std::cout << "Norm " << ebosResid.two_norm() << std::endl;
+            // std::cout << "Printing well residual in backward mode" << std::endl;
+            // const auto& well_container = wellModel().getWellContainer();
+            // for (const auto& well : well_container) {
+            //     std::cout << "********************************* " << std::endl;
+            //     std::cout << "Print residual for " << well->name() << std::endl;
+            //     auto reswell = well->getResWell();
+            //     std::cout << "Norm "<< reswell.two_norm() << std::endl;
+            // }
+            //wellModel().printResidual(std::cout);
+
+            const int nc = UgGridHelpers::numCells(grid_);
+            BVector lam(nc);// this should be the prevois adjoint vector
+            //BVector adjRhs(nc);// this should have contribution from prevois solve
+            // assume no contributions from pure reservoir
+            BVector adjRhs = rhs;           
+            wellModel().computeObj(dt);                 //prepare all values needed in the wellModels
+            wellModel().rhsAdjointRes(adjRhs);      //calculated reservoir right hand side based on the well contributions
+            wellModel().applyt(adjRhs);                 // add rhs from the schur complement of well equations
+            //
+            //NB we get the linerized version from the reservoir part to be modified
+            auto& ebosJac = ebosSimulator_.model().linearizer().matrix();
+            std::unique_ptr<Mat> adj_matrix_for_preconditioner;
+            if (param_.matrix_add_well_contributions_) {
+                wellModel().addWellContributions(ebosJac);
+            }
+            if ( param_.preconditioner_add_well_contributions_ &&
+                 ! param_.matrix_add_well_contributions_ ) {
+                adj_matrix_for_preconditioner.reset(new Mat(ebosJac));
+                wellModel().addWellContributions(*adj_matrix_for_preconditioner);
+            }
+
+            // then all well tings has to be done
+            // set initial guess
+            BVector x(nc);
+            if (param_.use_amgcl_ || param_.use_umfpack_) {
+                if (!param_.matrix_add_well_contributions_) {
+                    // This should be handled by combining interacting options into a single one eventually.
+                    OPM_THROW(std::runtime_error, "Cannot run with amgcl or UMFPACK without also using 'matrix_add_well_contributions=true'.");
+                }
 
+                // Create matrix for external linear solvers.
+                const auto& ebosJac_org = ebosSimulator_.model().linearizer().matrix();
+                auto ebosJac_trans = ebosJac_org;
+                auto adjRhs_cp = adjRhs;
```

Comment:
I am not against your comment, while this is not interesting at all. The mixing of the coding styles really makes the development suffering and a lot of time and energy are wasted during  development. Basically it is hurting everyone that wants to develop with OPM now. 

Again, I am really urging that we can have a coding style guide for the project. We might not agree with each other for a lot of things, let us make it through a democratic way with a simple majority vote. I will follow whatever involved in the guide. 

As long as we follow it in the new development, it will become better and better. Otherwise, there will be this kind of reviewing comments again and again, and it will not become better at all.

Reply:
> We might not agree with each other for a lot of things, let us make it through a democratic way with a simple majority vote.

While I'm not against a more binding coding style-guide per-se, I'm very much against applying the one-man-one-vote principle here, because I've been told that OPM is supposed to be a meritocratic and not necessarily a democratic project. I thus prefer something like a one-line-one-vote principle. Also, I'm of the opinion that the people who argue for a given change need willing to carry its implications, i.e., doing the work of making everything consistent in this case.

In the meantime, I propose that everyone writes his or her patches using whatever coding style they like and expected to get some mild nagging about style in the review. I think that there are currently much bigger hurdles to contributing.

> As long as we follow it in the new development, it will become better and better. Otherwise, there will be this kind of reviewing comments again and again, and it will not become better at all.

I agree that this would be reduce review noise somewhat, but I don't think comments about missing curly-braces or variable naming conventions are currently our biggest problem.

Reply:
> I'm very much against applying the one-man-one-vote principle here, because I've been told that OPM is supposed to be a meritocratic and not necessarily a democratic project. 

It is just a solution (easy and simple), while not necessary the solution will will use. I do not care about whatever solution we use. I will be very happy to have a guide that we do not think about what is the correct way to write a line of code anymore, and I do not want to see this kind of comment about coding style thing based on what file the code is in. 

> In the meantime, I propose that everyone writes his or her patches using whatever coding style they like and expected to get some mild nagging about style in the review. I think that there are currently much bigger hurdles to contributing.

Nagging is because of there is no guide people to follow. Nagging should not be necessary. 

> I agree that this would be reduce review noise somewhat, but I don't think comments about missing curly-braces or variable naming conventions are currently our biggest problem.

It is not our biggest problem for sure. But it is a problem has been there for long and will come back again and again and impacting development everyday.  For sure. the compilation time and memory usage of compilation are bigger problem than this one. They are causing more frustrations for sure.

Reply:
The current agreed-on approach is that we follow the style of the surrounding code. So the review comment from @andlaus about camelCase here is appropriate.

> In the meantime, I propose that everyone writes his or her patches using whatever coding style they like and expected to get some mild nagging about style in the review.

This, however, is *not* along the lines we have agreed. For now, write patches using the surrounding coding style, not your own...

If we at some point agree on a uniform coding style, it shall be one that is possible to automate using clang-format, all "conversion" work will be done in a day, and from that point there need be no such discussions in code reviews. I think clang-format is good and widespread enough that this could be done soon-ish.

> Also, I'm of the opinion that the people who argue for a given change need to be willing to carry its implications, i.e., doing the work of making everything consistent in this case.

With clang-format automated this would be easy.

Reply:
> This, however, is not along the lines we have agreed. For now, write patches using the surrounding coding style, not your own...

true. I wanted to say "try to use the style of the surrounding code, but if in doubt, don't try to be too perfect because mistakes happen. If they do, you'll hopefully get a friendly in the review.

> With clang-format automated this would be easy.

have you ever tried to use it? last time I did (which, admittedly is a few years back) I had to hack on it to make it approximately do what I wanted.